src/node_crypto.cc

```diff
@@ -3090,8 +3090,17 @@ void CipherBase::InitIv(const char* cipher_type,
   const int expected_iv_len = EVP_CIPHER_iv_length(cipher);
   const int mode = EVP_CIPHER_mode(cipher);
   const bool is_gcm_mode = (EVP_CIPH_GCM_MODE == mode);
+  const bool has_iv = iv_len >= 0;
 
-  if (is_gcm_mode == false && iv_len != expected_iv_len) {
+  // Throw if no IV was passed and the cipher requires an IV
+  if (!has_iv && expected_iv_len != 0) {
+    char msg[128];
+    snprintf(msg, sizeof(msg), "Missing IV for cipher %s", cipher_type);
+    return env()->ThrowError(msg);
+  }
+
+  // Throw if an IV was passed which does not match the cipher's fixed IV length
+  if (is_gcm_mode == false && has_iv && iv_len != expected_iv_len) {
```

Comment:
At least in theory we could end up calling `EVP_CIPHER_CTX_ctrl(EVP_CTRL_GCM_SET_IVLEN)` on line 3116 with a value of -1.

That's currently harmless, openssl returns an error, but I'm not sure if we have tests that exercise that particular code path.  Perhaps it's best to add a `CHECK(has_iv)` there.

Reply:
That case should be handled by `!has_iv && expected_iv_len != 0` (the `expected_iv_len` is 12 for GCM). Where should the `CHECK` be, after `EVP_CIPHER_CTX_ctrl(EVP_CTRL_GCM_SET_IVLEN)` fails?

Reply:
I'd do it right before the call to EVP_CIPHER_CTX_ctrl().